### PR TITLE
Enable loading ckpt for t0 finetuning

### PR DESCRIPTION
**Pull request by Muennighoff**

### CODE REVIEW

File: megatron/utils.py [hunk 1]
```diff
@@ -261,11 +261,18 @@ def get_packed_attention_mask(is_causal: bool, causal_mask: torch.Tensor, decode
...
+        decoder_is_inputs = torch.tensor([[1, 1, 1, 2, 2, 2, 0]])
+        segment_ids = torch.tensor([[1, 1, 0, 1, 1, 0, 0]])
+        causal_mask = torch.tril(torch.ones(att_mask_batch, seq_length, seq_length)).view(att_mask_batch, 1, seq_length, seq_length)
     """
 
     """Causal Inputs Mask:
-    mask = [[[[1, 1, 0, 0, 0, 0, 0],
-            [1, 1, 0, 0, 0, 0, 0],
+    mask = [[[[1, 1, 0, 1, 1, 0, 0],
+            [1, 1, 0, 1, 1, 0, 0],
```
Comment (thomasw21) [inline megatron/utils.py:275]: This is wrong I believe. I see no reason why this is like that. Is we support that the causal mask is casaul it should have zeros on the top right triangle.

File: megatron/utils.py [hunk 2]
```diff
@@ -299,7 +306,7 @@ def get_packed_attention_mask(is_causal: bool, causal_mask: torch.Tensor, decode
             [0, 0, 0, 1, 1, 1, 0],
             [0, 0, 0, 1, 1, 1, 0],
             [0, 0, 0, 1, 1, 1, 0],
-            [0, 0, 0, 0, 0, 0, 0]]]]
+            [0, 0, 0, 0, 0, 0, 1]]]]
```
Comment (thomasw21) [inline megatron/utils.py:309]: Nice catch.

File: megatron/utils.py [hunk 3]
```diff
@@ -311,13 +318,22 @@ def get_packed_attention_mask(is_causal: bool, causal_mask: torch.Tensor, decode
...
             [0, 0, 0, 0, 0, 0, 0]]]]
+    
+    If is_causal=True:
+    mask = [[[[1, 0, 0, 0, 0, 0, 0],
+            [1, 1, 0, 0, 0, 0, 0],
+            [1, 1, 1, 0, 0, 0, 0],
+            [0, 0, 0, 1, 0, 0, 0],
+            [0, 0, 0, 1, 1, 0, 0],
+            [0, 0, 0, 1, 1, 1, 0],
+            [0, 0, 0, 0, 0, 0, 0]]]]
```
Comment (thomasw21) [inline megatron/utils.py:329]:
```suggestion
            [0, 0, 0, 0, 0, 0, 1]]]]
```

File: megatron/utils.py (hunk 1 above)
Comment (Muennighoff) [reply megatron/utils.py:275]: I think it is correct.
decoder_is_inputs  (torch.tensor([[1, 1, 0, 1, 1, 0, 0]])) keeps these ones after its multiplication. In  the + with the causal_mask, which is just like an OR they are still kept.

It is not a problem because they are later removed when multiplying with the other masks.

File: megatron/utils.py (hunk 3 above)
Comment (Muennighoff) [reply megatron/utils.py:329]: I don't think there is a `1` , because the last row & column is 100% padding

Comment (thomasw21) [reply megatron/utils.py:329]: Hum I'm wondering if this doesn't screw something up. Essentially you're going to compute softmax on a row with only zeros ...

File: megatron/utils.py (hunk 1 above)
Comment (thomasw21) [reply megatron/utils.py:275]: Ah indeed you're right!

File: megatron/utils.py (hunk 3 above)
Comment (Muennighoff) [reply megatron/utils.py:329]: The last row & last col are the attention scores of the last token with respect to the last token. Since the last token is masked out in our loss_mask it doesn't matter I think.
Also it's a row with only -inf, no?

Comment (thomasw21) [reply megatron/utils.py:329]: No you compute softmax, what should be the result of the softmax of a row full of masked out values .... It feels like that would return lots of Nans.

File: finetune_t0_non_causal_decoder.py [hunk 4]
```diff
@@ -83,17 +98,20 @@ def get_batch_pipe(data):
...
     loss_on_targets_only = ~data_c["decoder_is_inputs"][:, 1:]
-    loss_on_non_pad_only = (tokens != tokenizer.pad)
+    loss_on_non_pad_only = (labels != tokenizer.pad)
     loss_mask *= loss_on_targets_only * loss_on_non_pad_only
 
     attention_mask = get_packed_attention_mask(
         # Run non-causal decoder
-        is_causal=False,
-        causal_mask=~(causal_mask.bool()),
+        is_causal=True,
```
[the remark this answers is not in the document]
Comment (Muennighoff) [reply finetune_t0_non_causal_decoder.py:92]: What about just `finetune_t0.py`?

File: megatron/utils.py (hunk 3 above)
Comment (Muennighoff) [reply megatron/utils.py:329]: Don't we fill it with `-inf`?
And the softmax of a row where all values are the same is just `1/n`, no? Where would it cause NaNs?

File: finetune_t0_non_causal_decoder.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (thomasw21) [reply finetune_t0_non_causal_decoder.py:92]: Right but do we hardcode this everytime? I'd rather have this one be the script for causal decoder.

File: megatron/utils.py (hunk 3 above)
Comment (thomasw21) [reply megatron/utils.py:329]: You can try writing a test but I would be pretty sure that the actual results are 0.

File: finetune_t0_non_causal_decoder.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Muennighoff) [reply finetune_t0_non_causal_decoder.py]: Added an argument `prefixlm`

File: finetune_t0.py [hunk 5]
```diff
@@ -83,17 +92,21 @@ def get_batch_pipe(data):
...
-        causal_mask=~(causal_mask.bool()),
+        is_causal=not(args.prefixlm),
+        causal_mask=~(causal_mask.bool()), # Turn back into tril being ones
         decoder_is_inputs=decoder_is_inputs.bool(),
         segment_ids=segment_ids.long(),
     )
 
+    if args.norm_target_loss:
+        loss_mask = loss_mask.view(-1)
+        loss_mask = fast_normalize(loss_mask)
```
Comment (thomasw21) [inline finetune_t0.py:108]: Maybe reshaping to the orignal structure is better API? It's better to bave the same shapes as label IMO (we still still do flatten everything)

File: tests/test_dataloaders.py [hunk 6]
```diff
@@ -241,7 +241,7 @@ def test_decoder_packed_mtf_dataloader(self):
                     last_padding_size = len([None for segment_id in items["decoder_segment_ids"][micro_batch_size - 1] if segment_id == 0])
 
 
-    def test_finetune_t0_non_causal_decoder_get_batch_pipe(self):
+    def test_finetune_t0_get_batch_pipe(self):
```
Comment (thomasw21) [inline tests/test_dataloaders.py:244]: Yeah let's make it so that the script is causal decoder specific. Let's figure out non causal decoder later on.

File: tools/preprocess_data.py [hunk 7]
```diff
@@ -117,6 +122,10 @@ def get_args():
                        help='Path to the BPE merge file (if necessary).')
     group.add_argument('--append-eod', action='store_true',
                        help='Append an <eod> token to the end of a document.')
+    group.add_argument('--append-bos', action='store_true',
+                       help='Append a bos token to the end of a document.')
+    group.add_argument('--prepend-space', action='store_true',
+                    help='Prepends a space to the beginning of a document')
```
Comment (thomasw21) [inline tools/preprocess_data.py:128]: Add a mention in which context it's useful, typically it is when you compute targets.

**Review by thomasw21 — CHANGES_REQUESTED** [review]
Nice work! Some things I think shouldn't be in this PR.

File: megatron/data/decoder_packed_mtf_dataset.py [hunk 8]
```diff
@@ -399,7 +399,7 @@ def _build_index_mappings(
     shuffle_idx_filename = _filename + '_decoder_packed_shuffle_idx.npy'
 
     # Build the indexed mapping if not exist.
-    if torch.distributed.get_rank() == 0:
+    if not torch.distributed.is_initialized() or torch.distributed.get_rank() == 0:
```
[the remark this answers is not in the document]
Comment (Muennighoff) [reply megatron/data/decoder_packed_mtf_dataset.py:402]: Afaik you added this code; I think it was for running tests or sth

[the remark this answers is not in the document]
Comment (thomasw21) [reply megatron/data/decoder_packed_mtf_dataset.py:402]: Arf probably because I wanted to use the data loader only ... Maybe let's remove for now because we should be assuming that torch distributed is always initialized at least in Meg-DS IMO.
